anydex/wallet/ethereum/eth_wallet.py

```diff
@@ -0,0 +1,238 @@
+        self._update_database(self.get_transactions())
+        pending_outgoing = self.get_outgoing_amount()
+        balance = {
+            'available': self.provider.get_balance(address) - pending_outgoing,
+            'pending': self.get_incoming_amount(),
+            'currency': 'ETH',
+            'precision': self.precision()
+        }
+        return succeed(balance)
+
+    def get_outgoing_amount(self):
+        """
+        Get the current amount of ethereum that we are sending, but is still unconfirmed.
+        :return: pending outgoing amount
+        """
+        outgoing = self._session.query(func.sum(Transaction.value)).filter(Transaction.is_pending.is_(True)).filter(
+            func.lower(Transaction.from_) == self.account.address.lower()).first()[0]
+        return outgoing if outgoing else 0
+
+    def get_incoming_amount(self):
+        """
+        Get the current amount of ethereum that is being sent to us, but is still unconfirmed.
+        :return: pending incoming amount
+        """
+        incoming = self._session.query(func.sum(Transaction.value)).filter(Transaction.is_pending.is_(True)).filter(
+            func.lower(Transaction.to) == self.account.address.lower()).first()[0]
+        return incoming if incoming else 0
+
+    async def transfer(self, amount, address) -> str:
+        """
+        Transfer Ethereum to another wallet.
+        If the amount exceeds the available balance, an `InsufficientFunds` exception is raised.
+
+        :param amount: the transfer amount
+        :param address: the receiver address
+        :return: transfer hash
+        """
+        balance = await self.get_balance()
+
+        if balance['available'] < int(amount):
+            raise InsufficientFunds('Insufficient funds')
+
+        self._logger.info(f'Creating Ethereum payment with amount {amount} to address {address}')
+
+        transaction = {
+            'from': self.get_address(),
+            'to': address,
+            'value': int(amount),
+            'nonce': self.provider.get_transaction_count(self.get_address()),
+            'gasPrice': self.provider.get_gas_price(),
+            'chainId': self.get_chain_id()
+        }
+
+        transaction['gas'] = self.provider.estimate_gas(transaction)
+        # submit to blockchain
+        signed = self.account.sign_transaction(transaction)
+        self.provider.submit_transaction(signed['rawTransaction'])
+
+        # add transaction to database
+        self._session.add(
+            Transaction(
+                from_=transaction['from'],
+                to=transaction['to'],
+                value=transaction['value'],
+                gas=transaction['gas'],
+                nonce=transaction['nonce'],
+                gas_price=transaction['gasPrice'],
+                hash=signed['hash'].hex(),
+                is_pending=True
+            )
+        )
+        self._session.commit()
+        return signed['hash'].hex()
+
+    def get_chain_id(self):
```

Comment:
What is the purpose of this method, and what exactly does the chain id entail? I'm noticing that it is hardcoded in all appearances.

Reply:
there are multiple ethereum chains, you have ethereum, ethereum classic and then their corresponding testnet. IIRC the chainid is there so that you cannot do replay attacks.
So you can't make transaction on ehtereum classic and then also broadcast it on the ethereum network.